### ISSUE TITLE
gotype installation failed: removed from x/tools

### ISSUE THREAD

**Issue by rasky** [golang/go#19516]
This commit removed `gotype` from `x/tools`:
https://github.com/golang/tools/commit/f5a6ee1ea9f7b3a91e3e70dc1b9706886d0e0ae3

After this commit, all existing references across the Internet to install `gotype` fails. Everywhere you read that, to install `gotype`, you need to run `go get -u golang.org/x/tools/cmd/gotype`, but this is now broken, for users of any Go version.

The first Google hit for `gotype` is now a 404:
https://github.com/golang/tools/tree/master/cmd/gotype

`gotype` is being used (in the current form) by a lot of IDE integrations and tooling. It's part of the daily workflow for many Go programmers. Can you please clarify why the command-line tool has been removed from the repo?

/cc @griesemer

**Comment by rasky** [comment]
For anybody hitting this, the current workaround is:

    $ cd $GOPATH/src/golang.org/x/tools
    $ git checkout f5a6ee1^
    $ go get golang.org/x/tools/cmd/gotype

**Comment by mvdan** [comment]
If I recall correctly, `gotype` was never meant for general use. After all, what use would it have as part of an IDE if the compiler already throws type errors?

I forget the details and can't find them now though, so leaving for @griesemer to clarify.

**Comment by mvdan** [comment]
Ah, found the comment: https://github.com/golang/go/issues/12703#issuecomment-191327190

Also see https://github.com/golang/go/issues/14727.

**Comment by rasky** [comment]
> gotype was never meant for general use

Then it shouldn't have been put into `x/tools/cmd`, where there are dozens of tools for people to use like `goimports`, `gorename`, `stringer`, etc. People now do use it and rely on it.

>After all, what use would it have as part of an IDE if the compiler already throws type errors?

It's being used to highlight source code errors without running a build, which is much slower and also produces artifacts (side-effects) which are annoying to handle.

In #12703, `gotype` was updated to use `go/types` and everything was fine that way. It's been now dropped in favor of a non-gogettable solution, for reasons that are unclear. This is effectively breaking tools, IDEs and workflows for may Go users. 

I understand that `x/tools` does not follow Go 1.0 guarantee, but between absolute backward compatibility and suddenly removing a very useful tool out of the blue, there has to be a middle way. At least, a deprecation period should be expected, once an alternative solution has been identified.

**Comment by myitcv** [comment]
@rasky I too used to use `gotype` for much the same purposes you describe. I reported one issue in https://github.com/golang/go/issues/18799 but there are others.

I'll leave @griesemer to comment on the removal of `gotype` (because I too noted the comments linked by @mvdan's when I started using it).

But perhaps there is scope here for the quick introduction of a replacement (not necessarily maintained by the core Go team), whether at `golang.org/x/tools/cmd/gotype` or elsewhere. @rasky - can you give details of which tools use/refer to `gotype` to your knowledge? And any accompanying details on how the tool was used?

That way we can fairly quickly resurrect `gotype` (or similar) to close the gap. 

@griesemer could a replacement exist within `golang.org/x/tools/...` if it were given a well-defined owner?

**Comment by griesemer** [comment]
The main reason for removing x/tools/cmd/gotype was that it wasn't the most up-to-date version: The latest version, and the one that is maintained (by me) is go/types/gotype.go. We really would prefer to not have multiple diverging sources.

I will just keep a 1:1 copy in x/tools/cmd for now.

**Comment by lurkermike** [comment]
I believe this may have introduced a build error in gotype:

```
> git pull
Already up-to-date.

> git rev-parse HEAD
381149a2d6e5d8f319ccf04bfefc71e03a78b868

> go build
# golang.org/x/tools/cmd/gotype
./gotype.go:285: undefined: types.SizesFor
```

**Comment by rasky** [comment]
Thanks @griesemer, very appreciated. Can you please clarify why it has been decided to move it as a non-gogettable copy within the standard library, rather than stay with similar tools in `x/tools`?

**Comment by griesemer** [comment]
@rasky It was not moved - it was eliminated from x/tools. There was an up-to-date version of gotype.go in the std library all along (ever since go/types moved to the std library). We really don't want to maintain two versions. Also, gotype was never really meant to be for general use; but I guess anything that's exposed one way or the other will be used, no matter what...

Unfortunately it looks like we have to maintain two versions now, because the latest version relies on code that's not available in versions before 1.8.
